iothub/service/src/Amqp/AmqpConnectionHandler.cs

```diff
@@ -0,0 +1,319 @@
+using System.Net.WebSockets;
+using System.Security.Authentication;
+using System.Security.Cryptography.X509Certificates;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.Amqp;
+using Microsoft.Azure.Amqp.Framing;
+using Microsoft.Azure.Amqp.Transport;
+
+namespace Microsoft.Azure.Devices.Amqp
+{
+    /// <summary>
+    /// Handles a single AMQP connection. Supports TCP and Websocket ports.
+    /// </summary>
+    /// <remarks>
+    /// This class intentionally abstracts away details about sessions and links for simplicity at the service client level.
+    /// </remarks>
+    internal class AmqpConnectionHandler
+    {
+        private AmqpConnection _connection;
+        private AmqpCbsSessionHandler _cbsSession;
+        private AmqpSessionHandler _workerSession;
+
+        private readonly bool _useWebSocketOnly;
+        private readonly IotHubServiceClientOptions _options;
+        private ClientWebSocketTransport _clientWebSocketTransport;
+        private IotHubConnectionProperties _credential;
+        private static readonly AmqpVersion s_amqpVersion_1_0_0 = new(1, 0, 0);
+
+        private EventHandler _connectionLossHandler;
+
+        private string _linkAddress;
+
+        // The current delivery tag. Increments after each send operation to give a unique value.
+        private int _sendingDeliveryTag;
+
+        public AmqpConnectionHandler(
+            IotHubConnectionProperties credential,
+            bool useWebSocketOnly,
+            string linkAddress,
+            IotHubServiceClientOptions options,
+            EventHandler connectionLossHandler,
+            Action<AmqpMessage> messageHandler = null)
+        {
+            _credential = credential;
+            _useWebSocketOnly = useWebSocketOnly;
+            _linkAddress = linkAddress;
+            _options = options;
+            _connectionLossHandler = connectionLossHandler;
+            _cbsSession = new AmqpCbsSessionHandler(_credential, connectionLossHandler);
+            _workerSession = new AmqpSessionHandler(linkAddress, connectionLossHandler, messageHandler);
+
+            _sendingDeliveryTag = 0;
+        }
+
+        /// <summary>
+        /// Opens the AMQP connection. This involves creating the needed TCP or Websocket transport and
+        /// then opening all the required sessions and links.
+        /// </summary>
+        /// <param name="cancellationToken">The cancellation token.</param>
+        public async Task OpenAsync(CancellationToken cancellationToken)
+        {
+            if (Logging.IsEnabled)
+                Logging.Enter(this, $"Opening amqp connection.");
+
+            try
+            {
+                AmqpSettings amqpSettings = CreateAmqpSettings();
+
+                TransportBase transport = null;
+                if (_useWebSocketOnly)
+                {
+                    // Try only AMQP transport over WebSocket
+                    transport = _clientWebSocketTransport =
```

Comment:
Why do we have both `transport` and `_clientWebSocketTransport` variables?

Reply:
I believe we can get rid of one of these, sure